### PR TITLE
"Merging branches 1"

### PR DESCRIPTION
**Pull request by rosheen33**

### CODE REVIEW

**Comment by alisufian** [comment]
Please remove `ass1.py~` file

**Comment by alisufian** [comment]
Use 4 spaces instead of tabs for indentation.

File: ass1.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+		
+	return None
+		
+	
+def main(*argv):
+	
+	total=0
+	
+	for i in argv:
+			total = total + 1
```
Comment (alisufian) [inline ass1.py]: Inconsistent spacing

**Comment by alisufian** [comment]
PR title should be better than "Merging branches 1"

File: ass1.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        for i in range(len(list)):
+
+            file = open(list[i], "r");
+            # Skip first two lines
+            next(file)
+            next(file)
+			
+            # Traverse all the lines of a file
+            for line in file:
+                currentline =line.split(",");
```
Comment (alisufian) [inline ass1.py]: wrong way to parse a csv file. use `csv.DictReader`

File: ass1.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
+import sys
+import re
+import os
+import glob
+
+
+def generate_report(reportNumber, directory):
+		
+    year=1996;
```
Comment (alisufian) [inline ass1.py]: No hardcoding.

File: ass1.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        print("-----------------------------------------------")
+
+    elif reportNumber== 3 :
+        print("Year                Date                Temp")
+        print("-----------------------------------------------")
+
+    # Traverse all files of all years
+    while (year<2012):
+        # list=glob.glob("/home/rosheen/Assignemnts/weatherdata/*"+str(year)+"*.txt");
+        list=glob.glob(directory+"/*"+str(year)+"*.txt");
```
Comment (alisufian) [inline ass1.py]: spacing issues

File: ass1.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+            print('{:4}'.format(year)+"             "+'{:5}'.format(maximunTemp)+"             "+'{:5}'.format(minimunTemp)+"            "+'{:5}'.format(maximumHumidity)+"            "+'{:5}'.format(minimumHumidity));			
+			
+        elif reportNumber == 2 :
+            # print(str(year)+"             "+hottet_day_date+"             "+str(maximunTemp))
+            print('{:4}'.format(year)+"             "+'{:10}'.format(hottet_day_date)+"             "+'{:5}'.format(maximunTemp))
+			
+        elif reportNumber == 3 :
+            # print(str(year)+"             "+coolest_day_date+"             "+str(minimunTemp))
+            print('{:4}'.format(year)+"             "+'{:10}'.format(coolest_day_date)+"             "+'{:5}'.format(minimunTemp))
+
```
Comment (alisufian) [inline ass1.py]: There should be no commented code.

File: ass1.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+		
+    elif(total == 2):
+        generate_report(argv[0], argv[1])
+		
+    return None
+		
+		
+# Call main function
+# main();		
+main(1,"/home/rosheen/Assignemnts/weatherdata");
```
Comment (alisufian) [inline ass1.py]: See Python documentation on how to setup `__main__`

File: ass1.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+        print("[Report Number]")
+        print("1 for Annual Max/Min Temperature")
+        print("2 for Hottest day of each year")
+        print("3 for coldest day of each year")
+
+        print("[data_dir]")
+        print("Directory containing weather data files")
+        sys.exit();
+		
+    elif(total == 2):
```
Comment (alisufian) [inline ass1.py]: parenthesis are not required.

File: ass1.py (hunk 5 above)
Comment (rosheen33) [inline ass1.py]: okay. I will remove it

File: ass1.py (hunk 6 above)
Comment (rosheen33) [inline ass1.py]: okay . i am on it .

File: ass1.py (hunk 7 above)
Comment (rosheen33) [inline ass1.py]: I just did it for my convenience , if you want me to remove it, I will remove it.

File: ass1.py (hunk 4 above)
Comment (rosheen33) [inline ass1.py]: Do u mean on line 15 and 19 ?

Comment (alisufian) [inline ass1.py]: 15, 19, 24, 26

File: ass1.py (hunk 7 above)
Comment (alisufian) [inline ass1.py]: yes lets remove those.

File: ass1.py (hunk 6 above)
Comment (alisufian) [inline ass1.py]: The arguments have to be taken from the command line. Use `argparse` to do that.

File: ass1.py (hunk 4 above)
Comment (rosheen33) [inline ass1.py]: kindly check now

File: ass1.py (hunk 6 above)
Comment (rosheen33) [inline ass1.py]: Kindly see now I have used sys.argv

File: ass1.py (hunk 7 above)
Comment (rosheen33) [inline ass1.py]: Removed

File: ass1.py (hunk 5 above)
Comment (rosheen33) [inline ass1.py]: Removed

File: ass1.py (hunk 2 above)
Comment (rosheen33) [inline ass1.py]: I am on it

Comment (rosheen33) [inline ass1.py]: i did not have a csv file. i had txt files in a zip folder

Comment (rosheen33) [inline ass1.py]: Oh i got it csv means "comma seprated"

Comment (rosheen33) [inline ass1.py]: I dont understand . Should not we use csv.dictReader when we have files with .csv extension ?

Comment (alisufian) [inline ass1.py]: its the data that matters the extension could be anything.

File: ass1.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    return None
+		
+		
+# Call main function
+if __name__ == "__main__":
+    main(*sys.argv)
+	
+	
+# main();		
+# main(1,"/home/rosheen/Assignemnts/weatherdata");
```
Comment (alisufian) [inline ass1.py]: Remove these.

File: ass1.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+		
+    elif(total >= 2):
+        generate_report(argv[1], argv[2])
+		
+    return None
+		
+		
+# Call main function
+if __name__ == "__main__":
+    main(*sys.argv)
```
Comment (alisufian) [inline ass1.py]: use `import argparse` for parsing arguments.

**Comment by alisufian** [comment]
Break down the code into functions. Create a function for each report.

File: ass1.py (hunk 8 above)
Comment (rosheen33) [inline ass1.py]: Done

File: ass1.py (hunk 2 above)
Comment (rosheen33) [inline ass1.py]: I have done it . kindly see it

On 17 August 2015 at 18:07, Ali Sufian <notifications@github.com> wrote:

> In ass1.py
> <https://github.com/arbisoft/the-lab/pull/43#discussion_r37185180>:
>
> > +        maximumHumidity=0
> > +        minimumHumidity=10000
> > +        hottet_day_date=""
> > +        coolest_day_date=""
> > +		
> > +        # Traverse all the files of a specific year
> > +        for i in range(len(list)):
> > +
> > +            file = open(list[i], "r");
> > +            # Skip first two lines
> > +            next(file)
> > +            next(file)
> > +			
> > +            # Traverse all the lines of a file
> > +            for line in file:
> > +                currentline =line.split(",");
>
> its the data that matters the extension could be anything.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/arbisoft/the-lab/pull/43/files#r37185180>.
>



-- 

*Rosheen Pervaiz*

File: ass1.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
+import sys
+import glob
+import csv
+
+
+def generate_report(reportNumber, directory):
+		
+    # Read first file to get a year with lowest year
+    list = glob.glob(directory+ "/*.txt")
```
Comment (alisufian) [inline ass1.py]: incorrect spacing.
`list` is a built-in so do not use that as a variable name.

File: ass1.py [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+
+    elif reportNumber == '3' :
+        print "Year                Date                Temp"
+        print "-----------------------------------------------"
+
+    #with open('/home/rosheen/Assignemnts/weatherdata/lahore_weather_1996_Dec.txt', 'rb') as csvfile:
+     #   spamreader = csv.reader(csvfile, delimiter=' ', quotechar='|')
+      #  for row in spamreader:
+       #         print ', '.join(row)
+    # Traverse all files of all years
```
Comment (alisufian) [inline ass1.py]: Please stop checking-in commented out code.

File: ass1.py [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+    while (year < 2012):
```
Comment (alisufian) [inline ass1.py]: no hardcoding

File: ass1.py [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+        minimunTemp = 10000
+        maximumHumidity = 0
+        minimumHumidity = 10000
+        hottet_day_date = ""
+        coolest_day_date = ""
+		
+        # Traverse all the files of a specific year
+        for i in range(len(list)):						
+            file = open(list[i], "r");
+            with open(list[i], 'rb') as csvfile:
```
Comment (alisufian) [inline ass1.py]: why is the file opened twice?

File: ass1.py [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+                spamreader = csv.reader(csvfile, delimiter=' ', quotechar='|')
```
Comment (alisufian) [inline ass1.py]: `spamreader`? copy paste with care.

File: ass1.py [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+                # Skip first two lines
+                spamreader.next()
+                spamreader.next()
+               
+                # Traverse all the lines of a file
+                for line in spamreader:
+                    str1 = line[0]
+                    line = str1
+                    currentline = line.split(",");
```
Comment (alisufian) [inline ass1.py]: If you are still tokenizing the line on comma then there is no point in using csv reader.

File: ass1.py [hunk 16]
```diff
@@ -0,0 +1,210 @@
...
+			
+        maximunTemp = 0
+        minimunTemp = 10000
+        maximumHumidity = 0
+        minimumHumidity = 10000
+		
+        # Traverse all the files of a specific year
+        for i in range(len(list1)):						
+            with open(list1[i], 'rb') as csvfile:
+                reader1 = csv.reader(csvfile, delimiter = ',')
```
Comment (alisufian) [inline ass1.py]: delimiter is not required.

File: ass1.py [hunk 17]
```diff
@@ -0,0 +1,210 @@
...
+                # Skip first two lines
+                reader1.next()
+                reader1.next()
+               
+                # Traverse all the lines of a file
+                for line in reader1:
+                    currentline = line
+		    				
+                    # Maximun temperature	
+                    if((len(currentline) > 1) and currentline[1] != '' and (int(currentline[1]) >= maximunTemp)):
```
Comment (alisufian) [inline ass1.py]: if you use DictReader you wont have to index into the list.

File: ass1.py (hunk 16 above)
Comment (rosheen33) [inline ass1.py]: okay . I will remove it

**Comment by alisufian** [comment]
The overall approach needs to be improved. Suppose all reports have to be run, then the file should be read once and reports generated. Each report should not be reading and parsing the file.

File: ass1.py (hunk 17 above)
Comment (rosheen33) [inline ass1.py]: I tried to use it first but it did not work that's why I use reader only. But I will look closer on it now

**Comment by alisufian** [comment]
And you are missing the basic component of a computer program: a "Data Structure" to hold the data you are processing.

**Comment by rosheen33** [comment]
okay . I will work on that

On 18 August 2015 at 16:32, Ali Sufian <notifications@github.com> wrote:

> The overall approach needs to be improved. Suppose all reports have to be
> run, then the file should be read once and reports generated. Each report
> should not be reading and parsing the file.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/arbisoft/the-lab/pull/43#issuecomment-132181612>.
>



-- 

*Rosheen Pervaiz*

**Comment by rosheen33** [comment]
Can I make a separate class ?

On 18 August 2015 at 16:35, Ali Sufian <notifications@github.com> wrote:

> And you are missing the basic component of a computer program: a "Data
> Structure" to hold the data you are processing.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/arbisoft/the-lab/pull/43#issuecomment-132182107>.
>



-- 

*Rosheen Pervaiz*

**Comment by alisufian** [comment]
Yes. (There isn't anything else that you can do :) )

File: weather.py [hunk 18]
```diff
@@ -0,0 +1,14 @@
...
+        self.max_temp = max_temp
+        self.min_temp = min_temp
+        self.max_hum = max_hum
+        self.min_hum = min_hum
+        self.hot_day = hot_day
+        self.cool_day = cool_day
+        self.year = year
+    
+
+    def displayEmployee(self):
```
Comment (alisufian) [inline weather.py]: Please at least change the name when copy pasting code.

File: ass1.py [hunk 19]
```diff
@@ -0,0 +1,181 @@
...
+				
+                    # Maximum Humidity
+                    if((len(currentline) > 1) and currentline[7] != '' and (int(currentline[7]) >= maximumHumidity)):
+                        maximumHumidity = int(currentline[7])
+
+        w1= Weather(maximunTemp, minimunTemp, maximumHumidity, minimumHumidity, hottet_day_date, coolest_day_date, year)
+        list2.append(w1)
+			
+        year = year + 1
+
```
Comment (alisufian) [inline ass1.py]: The logic of calculating each report must remain in its respective function. No point in creating functions just to print results.

File: ass1.py [hunk 20]
```diff
@@ -0,0 +1,219 @@
...
+        print "[data_dir]" 
+        print "Directory containing weather data files"
+        sys.exit();
+		
+    elif(argv.filename != '' and argv.num != 0):
+        directory = argv.filename
+        reportNumber = argv.num
+		
+        list2 = read_files(directory)
```
Comment (alisufian) [inline ass1.py]: You need to be a little more creative with variable names.

Comment (rosheen33) [inline ass1.py]: okay

Comment (rosheen33) [inline ass1.py]: changed the names of variables list1   list2

On 19 August 2015 at 12:55, Ali Sufian <notifications@github.com> wrote:

> In ass1.py
> <https://github.com/arbisoft/the-lab/pull/43#discussion_r37388175>:
>
> > +        print "Usage: weatherman <report Number> <data_dir>"
> > +
> > +        print "[Report Number]"
> > +        print "1 for Annual Max/Min Temperature"
> > +        print "2 for Hottest day of each year"
> > +        print "3 for coldest day of each year"
> > +
> > +        print "[data_dir]"
> > +        print "Directory containing weather data files"
> > +        sys.exit();
> > +		
> > +    elif(argv.filename != '' and argv.num != 0):
> > +        directory = argv.filename
> > +        reportNumber = argv.num
> > +		
> > +        list2 = read_files(directory)
>
> You need to be a little more creative with variable names.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/arbisoft/the-lab/pull/43/files#r37388175>.
>



-- 

*Rosheen Pervaiz*

**Comment by alisufian** [comment]
Much better, this can still be improved but I guess we've already had too many iterations. You can take a look here for improvement.
https://github.com/arbisoft/the-lab/pull/38/files

**Comment by rosheen33** [comment]
Should I improve my code by seeing the above code ?

On 19 August 2015 at 16:30, Ali Sufian <notifications@github.com> wrote:

> Much better, this can still be improved but I guess we've already had too
> many iterations. You can take a look here for improvement.
> https://github.com/arbisoft/the-lab/pull/38/files
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/arbisoft/the-lab/pull/43#issuecomment-132554197>.
>



-- 

*Rosheen Pervaiz*

**Comment by alisufian** [comment]
Not really required but you can to get the hang of it.

**Pull request merged**